db/import_test.go

```diff
@@ -0,0 +1,197 @@
+package db
+
+import (
+	"testing"
+	"time"
+
+	"fmt"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbaselabs/go.assert"
+	"log"
+)
+
+// There are additional tests that exercise the import functionality in rest/import_test.go
+
+// 1. Write a doc to the bucket
+// 2. Build params to migrateMeta (existing doc, raw doc, cas.. sgbucket docs)
+// 3. Update doc in the bucket with new expiry
+// 4. Call migrateMeta with stale args that have old expiry
+// 5. Assert that migrateMeta does the right thing and respects the new expiry value
+//
+// See SG PR #3109 for more details on motivation for this test
+func TestMigrateMetadata(t *testing.T) {
+
+	if !base.TestUseXattrs() {
+		t.Skip("This test only works with XATTRS enabled")
+	}
+
+	base.EnableLogKey("Migrate")
+	base.EnableLogKey("Import+")
+
+	db, testBucket := setupTestDB(t)
+	defer testBucket.Close()
+	defer tearDownTestDB(t, db)
+
+	key := "TestMigrateMetadata"
+	bodyBytes := rawDocWithSyncMeta()
+	body := Body{}
+	err := body.Unmarshal(bodyBytes)
+	assertNoError(t, err, "Error unmarshalling body")
+
+	// Create via the SDK with sync metadata intact
+	expirySeconds := time.Second * 30
+	syncMetaExpiry := time.Now().Add(expirySeconds)
+	_, err = testBucket.Bucket.Add(key, uint32(syncMetaExpiry.Unix()), bodyBytes)
+	assertNoError(t, err, "Error writing doc w/ expiry")
+
+	// Get the existing bucket doc
+	_, existingBucketDoc, err := db.GetDocWithXattr(key, DocUnmarshalAll)
+
+	// Set the expiry value to a stale value (it's about to be stale, since below it will get updated to a later value)
+	existingBucketDoc.Expiry = uint32(syncMetaExpiry.Unix())
+
+	// Update doc in the bucket with new expiry
+	laterExpirySeconds := time.Second * 60
+	laterSyncMetaExpiry := time.Now().Add(laterExpirySeconds)
+	updateCallbackFn := func(current []byte) (updated []byte, expiry *uint32, err error) {
+		// This update function will not be "competing" with other updates, so it doesn't need
+		// to handle being called back multiple times or performing any merging with existing values.
+		exp := uint32(laterSyncMetaExpiry.Unix())
+		return bodyBytes, &exp, nil
+	}
+	testBucket.Bucket.Update(
+		key,
+		uint32(laterSyncMetaExpiry.Unix()), // it's a bit confusing why the expiry needs to be passed here AND via the callback fn
+		updateCallbackFn,
+	)
+
+	// Call migrateMeta with stale args that have old stale expiry
+	_, _, err = db.migrateMetadata(
+		key,
+		body,
+		existingBucketDoc,
+	)
+	assert.True(t, err != nil)
+	assert.True(t, err == base.ErrCasFailureShouldRetry)
+
+}
+
+func TestImportDocWithStaleDoc(t *testing.T) {
```

Comment:
This test doesn't looks like it handles the case of a document that already has the SG xattr, and subsequently has an SDK write.  Based on the test name (TestImportDoc...), I expected that here.

rawDocWithSyncMeta is going to trigger migration, and not import.

Reply:
I renamed the test and added docs on the scenarios that it tests:

```
// This invokes db.importDoc() with two different scenarios:
//
// Scenario 1: normal import
//
// 1. Write doc via SDK that is a pure KV doc, no sync metadata `{"key": "val"}` and with expiry value expiryA
// 2. Perform an update via SDK to update the expiry to expiry value expiryB
// 3. Invoke db.importDoc() and pass it the stale doc from step 1, that has expiryA
// 4. Do a get on the doc and verify that it has the later expiry value expiryB, which verifies it did a CAS retry
//
// Scenario 2: import with migration
//
// - Same as scenario 1, except that in step 1 it writes a doc with sync metadata, so that it excercises the migration code
//
func TestImportWithStaleBucketDocCorrectExpiry(t *testing.T) {

```

Also, I verified that it adds [code coverage](http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-coverage/25/artifact/cover_sg_merged.html#file50) for the case that was discussed in last PR review that was previously missing code coverage:

<img width="1115" alt="screen shot 2017-12-06 at 5 50 07 pm" src="https://user-images.githubusercontent.com/296876/33694659-33316460-daae-11e7-903e-c8e7cd86f1c3.png">